packages/flutter/lib/src/painting/text_style.dart

```diff
@@ -1567,3 +1577,83 @@ class TextStyle with Diagnosticable {
     styles.add(EnumProperty<TextOverflow>('${prefix}overflow', overflow, defaultValue: null));
   }
 }
+
+/// Interpolate between two lists of [FontVariation] objects.
+///
+/// Variations are paired by axis, and interpolated using [FontVariation.lerp].
+///
+/// This algorithm is O(N), but the constant factor varies based on the input.
+///
+/// It is fastest when the lists contain the same axes ([FontVariation.axis]) in
+/// the same order. The result is again in the same order, and no attempt is
+/// made to detect or remove duplicates in this process.
+///
+/// When the order of items in the two input lists vary, the constant factor
+/// increases substantially, as it involves creating two maps and a set,
+/// inserting every variation in both lists into the maps and the set, and then
+/// iterating over them to recreate the list.
+///
+/// In this case, the resulting order is arbitrary. Duplicates are dropped; in
+/// each list, the last [FontVariation] for any particular axis is the one used
+/// to compute the value for that axis. Values that only appear on one side are
+/// interpolated using [FontVariation.lerp] against a null value, and resulting
+/// null values are omitted from the resulting list.
+///
+/// When the lists begin with matching pairs of axes, the fast algorithm is used
+/// up to the point where the lists diverge, then the more expensive algorithm
+/// is used.
+///
+/// See also:
+///
+///  * [TextStyle.lerp], which uses this function to handle
+///    [TextStyle.fontVariations].
+List<FontVariation>? lerpFontVariations(List<FontVariation>? a, List<FontVariation>? b, double t) {
+  if (t == 0.0) {
+    return a;
+  }
+  if (t == 1.0) {
+    return b;
+  }
+  if (a == null || a.isEmpty || b == null || b.isEmpty) {
+    // If one side is empty, that means anything on the other
+    // side will use the null-to-something lerp, which is to
+    // say, a step function at t=0.5.
+    return t < 0.5 ? a : b;
+  }
+  assert(a.isNotEmpty && b.isNotEmpty);
+  final List<FontVariation> result = <FontVariation>[];
+  // First, try the efficient O(N) solution in the event that
+  // the lists are compatible.
+  int index = 0;
+  final int minLength = a.length < b.length ? a.length : b.length;
+  for (; index < minLength; index += 1) {
+    if (a[index].axis != b[index].axis) {
+      // The lists aren't compatible.
+      break;
+    }
+    result.add(FontVariation.lerp(a[index], b[index], t)!);
+  }
+  final int maxLength = a.length > b.length ? a.length : b.length;
+  if (index < maxLength) {
+    // If we get here, we have found some case where we cannot
+    // use the efficient approach.
+    final Set<String> axes = HashSet<String>();
+    final Map<String, FontVariation> aVariations = HashMap<String, FontVariation>();
```

Comment:
Looking at the spec it looks like currently there're only 5 registered axes. Wouldn't it be faster to use regular element look up in a nested loop (or even sort the 2 lists), than doing the allocations ?

Reply:
Allocations are very cheap in Dart (it just bumps a pointer on the new generation heap).
I think if we expect this codepath to be used frequently enough that we care about performance, we should probably benchmark the various options, though. I don't know what the optimal path is. (I spent some time trying to figure it out, but nothing obvious came to mind.) In practice, I expect most cases will be either empty on one side and not the other (which has a fast path already), or the same order on both sides (which also has a fast path). There's not much benefit to lerping between two different lists (they would just step-function at t=0.5), and if the lists are the same, then it'd be much better for the values to be sorted by the developer than here.

Reply:
yeah it's not trivial (hence the paragraph in the documentation about it). I agree that the constant factors are probably much more important here than the scaling factor. I'll add some more text to that effect to the docs.